pkg/cache/cache.go

```diff
@@ -181,17 +223,70 @@ func New(defaultAudience, prefix string, defaultRegionalSTS bool, defaultTokenEx
 						return
 					}
 				}
-				c.pop(sa.Name, sa.Namespace)
+				c.popSA(sa.Name, sa.Namespace)
 			},
 			UpdateFunc: func(oldObj, newObj interface{}) {
 				sa := newObj.(*v1.ServiceAccount)
 				c.addSA(sa)
 			},
 		},
 	)
+	if cmInformer != nil {
+		cmInformer.Informer().AddEventHandler(
+			cache.ResourceEventHandlerFuncs{
+				AddFunc: func(obj interface{}) {
+					err := c.populateCacheFromCM(nil, obj.(*v1.ConfigMap))
+					if err != nil {
+						utilruntime.HandleError(err)
+					}
+				},
+				UpdateFunc: func(oldObj, newObj interface{}) {
+					err := c.populateCacheFromCM(oldObj.(*v1.ConfigMap), newObj.(*v1.ConfigMap))
+					if err != nil {
+						utilruntime.HandleError(err)
+					}
+				},
+			},
+		)
+	}
 	return c
 }
 
+func (c *serviceAccountCache) populateCacheFromCM(oldCM, newCM *v1.ConfigMap) error {
```

Comment:
I don't think it is deterministic the end result of the cache if two Updates or an Update and an Add (at the 60s resync period) happen at the same time, it depends on the order of pushes/pops. Of course then the cache should get fixed again at the next 60s resync so it's not a big issue but a queue would avoid it.

Reply:
Doesn't the informer framework use an underlying channel? So they should run sequentially, if I understand this correctly, not concurrently (no go routines).

Reply:
no you're right, the informer events are ordered and blocking, not sure where i got the impression otherwise but educated myself :woman_facepalming: https://github.com/kubernetes/client-go/blob/v0.23.0/tools/cache/shared_informer.go#L34